doc/diagnostics/AddingTracepoints.md

```diff
@@ -0,0 +1,417 @@
+
+Other parameters:
+
+`Obsolete` Guards against usage of the old tracepoint in the current code stream, but keeps it available for the formatter to use when processing trace from older releases.
+
+`Test` See the Advanced use #1 section.
+
+
+### Tracepoint Format Specifiers
+
+| Specifier | Description |
+| --------- | ----------- |
+| %d | 32-bit integer, formatted as signed decimal |
+| %u | 32-bit integer, formatted as unsigned decimal |
+| %lld | 64-bit integer, formatted as signed decimal |
+| %llu | 64-bit integer, formatted as unsigned decimal |
+| %x | 32-bit integer, formatted as hexadecimal |
+| %llx | 64-bit integer, formatted as hexadecimal |
+| %zu | platform-dependent size_t (equivalent to J9 UDATA), formatted as unsigned decimal |
+| %zd | platform-dependent size_t (equivalent to J9 IDATA), formatted as signed decimal |
+| %zx | platform-dependent size_t (equivalent to J9 UDATA), formatted as hexadecimal |
+| %f | float or double, formatted in fixed-point notation |
+| %p | 32-bit pointer in 32-bit builds, 64-bit pointer on 64-bit builds, formatted as hexadecimal |
+| %s | pointer to null-terminated character array, formatted as character string |
+| %.*s | 32-bit integer followed by pointer to character array, formatted as character string of specified length |
+
+
+The `ut_xxx.h` and `ut_xxx.c` files are refreshed by the tracegen step of the build. `ut_xxx.h` contains the macro you need.
+
+The .tdf entry above generates the following macro in `ut_j9vm.h`:
+
+```
+#if UT_TRACE_OVERHEAD >= 1
+#define Trc_VM_getMethodOrFieldID_dereferencedClass(thr, P1, P2, P3) do { /* tracepoint name: j9vm.123 */ \
+	if ((unsigned char) j9vm_UtActive[123] != 0){ \
+		j9vm_UtModuleInfo.intf->Trace(UT_THREAD(thr), &j9vm_UtModuleInfo, ((123u << 8) | j9vm_UtActive[123]), "\6\12\377", P1, P2, P3);} \
+	} while(0)
+#else
+#define Trc_VM_getMethodOrFieldID_dereferencedClass(thr, P1, P2, P3)   /* tracepoint name: j9vm.123 */
+#endif
+```
+
+Note the number of the new tracepoint is detailed in the comment - this example is `j9vm.123`.
+
+Tracegen also updates `lib/J9TraceFormat.dat`, which contains the tracepoint information used by the trace formatter (the Template string from your new tracepoint goes in there).
+
+3) Nearly there - add the tracepoint macro to your source code:
+
+```
+Trc_VM_getMethodOrFieldID_dereferencedClass(vmThread, clazz,
+        (UDATA)J9UTF8_LENGTH(J9ROMCLASS_CLASSNAME(clazz->romClass)),
```

Comment:
The cast to `(UDATA)` should be `(U_32)` (the table above says the length is a 32-bit integer).